vowpalwabbit/v_array.h

```diff
@@ -39,25 +54,70 @@ struct v_array
   T* end_array;
   size_t erase_count;
 
-  using iterator = T*;
+  using value_type = T;
+  using reference = value_type&;
+  using const_reference = const value_type&;
+  using pointer = value_type*;
+  using const_pointer = const value_type*;
+  using iterator = value_type*;
+  using const_iterator = const value_type*;
 
   // enable C++ 11 for loops
-  inline T*& begin() { return _begin; }
-  inline T*& end() { return _end; }
-
-  inline const T* begin() const { return _begin; }
-  inline const T* end() const { return _end; }
-
-  inline T* cbegin() const { return _begin; }
-  inline T* cend() const { return _end; }
-
-  // TODO uncomment this vv
-  // v_array cannot have a user-defined constructor, because it participates in various unions.
-  // union members cannot have user-defined constructors.
-  // v_array() : _begin(nullptr), _end(nullptr), end_array(nullptr), erase_count(0) {}
-  // ~v_array() {
-  //  delete_v();
-  // }
+  inline iterator& begin() { return _begin; }
+  inline iterator& end() { return _end; }
+
+  inline const_iterator begin() const { return _begin; }
+  inline const_iterator end() const { return _end; }
+
+  inline const_iterator cbegin() const { return _begin; }
+  inline const_iterator cend() const { return _end; }
+
+  v_array() noexcept : _begin(nullptr), _end(nullptr), end_array(nullptr), erase_count(0) {}
+  ~v_array() { delete_v_array(); }
+
+  v_array(v_array<T>&& other) noexcept
+  {
+    erase_count = 0;
+    _begin = nullptr;
+    _end = nullptr;
+    end_array = nullptr;
+
+    std::swap(_begin, other._begin);
+    std::swap(_end, other._end);
+    std::swap(end_array, other.end_array);
+    std::swap(erase_count, other.erase_count);
+  }
+
+  v_array<T>& operator=(v_array<T>&& other) noexcept
+  {
+    delete_v_array();
```

Comment:
shouldn't we check for
`if (this == &other)` here too?

Reply:
I believe the answer to this is no https://stackoverflow.com/questions/9322174/move-assignment-operator-and-if-this-rhs